R/mock-session.R

```diff
@@ -0,0 +1,283 @@
+      return(400)
+    } else if (att == "hidden") {
+      return(FALSE)
+    }
+  }
+  warning("Unexpected clientdata attribute accessed: ", name)
+  return(NULL)
+}
+
+#' @noRd
+#' @export
+`[[.mockclientdata` <- `$.mockclientdata`
+
+#' @noRd
+#' @export
+`[.mockclientdata` <- function(values, name) {
+  stop("Single-bracket indexing of mockclientdata is not allowed.")
+}
+
+#' @include timer.R
+MockShinySession <- R6Class(
+  'MockShinySession',
+  portable = FALSE,
+  class = FALSE,
+  public = list(
+    env = NULL,
+    # Needed for rendering HTML (i.e. renderUI)
+    singletons = character(0),
+    # Define a mock client data that always returns a size for plots
+    clientData = structure(list(), class="mockclientdata"),
+    reactlog = function(logEntry){},
+    incrementBusyCount = function(){},
+    output = NULL,
+    input = NULL,
+    userData = NULL,
+
+    initialize = function() {
+      private$.input <- ReactiveValues$new(dedupe = FALSE, label = "input")
+      private$flushCBs <- Callbacks$new()
+      private$flushedCBs <- Callbacks$new()
+      private$endedCBs <- Callbacks$new()
+      private$timer <- MockableTimerCallbacks$new()
+
+      self$userData <- new.env(parent=emptyenv())
+
+      # create output
+      out <- .createOutputWriter(self)
+      class(out) <- "shinyoutput"
+      self$output <- out
+
+      # Create a read-only copy of the inputs reactive.
+      self$input <- .createReactiveValues(private$.input, readonly = TRUE)
+    },
+    onFlush = function(fun, once) {
+      if (!isTRUE(once)) {
+        return(private$flushCBs$register(fun))
+      } else {
+        dereg <- private$flushCBs$register(function() {
+          dereg()
+          fun()
+        })
+        return(dereg)
+      }
+    },
+    onFlushed = function(fun, once) {
+      if (!isTRUE(once)) {
+        return(private$flushedCBs$register(fun))
+      } else {
+        dereg <- private$flushedCBs$register(function() {
+          dereg()
+          fun()
+        })
+        return(dereg)
+      }
+    },
+    onEnded = function(sessionEndedCallback) {
+      private$endedCBs$register(sessionEndedCallback)
+    },
+
+    isEnded = function(){ private$closed },
+    isClosed = function(){ private$closed },
+    close = function(){ private$closed <- TRUE },
+
+    #FIXME: this is wrong. Will need to be more complex.
+    cycleStartAction = function(callback){ callback() },
+
+    # Needed for image rendering. Base64-encode the given file.
+    fileUrl = function(name, file, contentType='application/octet-stream') {
+      bytes <- file.info(file)$size
+      if (is.na(bytes))
+        return(NULL)
+
+      fileData <- readBin(file, 'raw', n=bytes)
+      b64 <- rawToBase64(fileData)
+      return(paste('data:', contentType, ';base64,', b64, sep=''))
+    },
+
+    setInputs = function(...) {
+      vals <- list(...)
+      # TODO: is there really not a way to access `names` from inside an lapply?
+      lapply(names(vals), function(k){
```

Comment:
The way this is usually done is:

```R
mapply(names(vals), vals, function(name, value) {
  private$.input$set(name, value)
})
```

If the return value matters, you may twiddle with the `SIMPLIFY` and `USE.NAMES` arguments, but I don't think that's the case here.